### ISSUE TITLE
Getting permission denied when attempting a build with SELinux Enforcing on Fedora 22

### ISSUE THREAD

**Issue by maxamillion** [projectatomic/atomic-reactor#206]
When running the test build command from the `README.md` it fails when SELinux is Enforcing on a Fedora 22 machine (this is however successful if SELinux is Permissive):
```
# atomic-reactor build git --method hostdocker --build-image buildroot --image test-image --uri "https://github.com/TomasTomecek/docker-hello-world.git"
2015-07-01 18:29:17,554 - atomic_reactor.outer - INFO - building image 'test-image'
2015-07-01 18:29:17,559 - atomic_reactor.core - INFO - building image 'buildroot' in container using docker from host
2015-07-01 18:29:17,559 - atomic_reactor.core - INFO - checking whether image 'buildroot' exists
2015-07-01 18:29:17,574 - atomic_reactor.core - INFO - creating container from image 'buildroot' and running it
2015-07-01 18:29:18,247 - atomic_reactor.core - INFO - getting stdout of container '822a9a8e498302ac2f7cb111f1a5f248440810b4058a4d29de0b9b54470f7f76'
2015-07-01 18:29:18,613 - atomic_reactor.util - INFO - wait_for_command
2015-07-01 18:29:21,096 - atomic_reactor.util - INFO - no more logs
2015-07-01 18:29:21,097 - atomic_reactor.core - INFO - waiting for container '822a9a8e498302ac2f7cb111f1a5f248440810b4058a4d29de0b9b54470f7f76' to finish
2015-07-01 18:29:21,322 - atomic_reactor.core - INFO - getting stdout of container '822a9a8e498302ac2f7cb111f1a5f248440810b4058a4d29de0b9b54470f7f76'
2015-07-01 18:29:21,353 - atomic_reactor - ERROR - build failed
```

I then look at the `docker logs` output of the container and see the permission denied:
```
2015-07-01 18:17:13,567 - atomic_reactor.inner - DEBUG - getting build json from input path
2015-07-01 18:17:13,567 - atomic_reactor.plugin - DEBUG - loading plugins from dir '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins'
2015-07-01 18:17:13,569 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_inject_yum_repo.py'
2015-07-01 18:17:13,570 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_push_to_pulp.py'
2015-07-01 18:17:13,572 - atomic_reactor.plugin - WARNING - can't load module '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_push_to_pulp.py': ImportError('No module named dockpulp',)
2015-07-01 18:17:13,572 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/input_osv3.py'
2015-07-01 18:17:13,573 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_store_metadata_in_osv3.py'
2015-07-01 18:17:13,574 - atomic_reactor.plugin - WARNING - can't load module '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_store_metadata_in_osv3.py': ImportError('No module named osbs.api',)
2015-07-01 18:17:13,574 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_add_yum_repo.py'
2015-07-01 18:17:13,575 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/__init__.py'
2015-07-01 18:17:13,575 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_store_logs_to_file.py'
2015-07-01 18:17:13,576 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_remove_built_image.py'
2015-07-01 18:17:13,577 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_koji.py'
2015-07-01 18:17:15,116 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_add_dockerfile.py'
2015-07-01 18:17:15,116 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_rpmqa.py'
2015-07-01 18:17:15,117 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_change_source_registry.py'
2015-07-01 18:17:15,117 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_return_dockerfile.py'
2015-07-01 18:17:15,117 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_add_yum_repo_by_url.py'
2015-07-01 18:17:15,118 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_tag_and_push.py'
2015-07-01 18:17:15,118 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_cp_dockerfile.py'
2015-07-01 18:17:15,118 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/input_path.py'
2015-07-01 18:17:15,119 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_pyrpkg_fetch_artefacts.py'
2015-07-01 18:17:15,119 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/prepub_tests_for_image.py'
2015-07-01 18:17:15,120 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/input_env.py'
2015-07-01 18:17:15,120 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_change_from_in_df.py'
2015-07-01 18:17:15,120 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/prepub_squash.py'
2015-07-01 18:17:15,121 - atomic_reactor.plugin - WARNING - can't load module '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/prepub_squash.py': ImportError('No module named docker_scripts.squash',)
2015-07-01 18:17:15,121 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_cp_built_image_to_nfs.py'
2015-07-01 18:17:15,122 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_assert_labels.py'
2015-07-01 18:17:15,122 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/pre_add_labels_in_df.py'
2015-07-01 18:17:15,122 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/plugins/post_tag_by_labels.py'
2015-07-01 18:17:15,123 - atomic_reactor.plugin - DEBUG - running plugin 'path'
2015-07-01 18:17:15,123 - atomic_reactor.inner - DEBUG - build json: {u'parent_registry_insecure': False, u'target_registries_insecure': False, u'dont_pull_base_image': False, u'image': u'test-image', u'source': {u'provider': u'git', u'provider_params': {u'git_commit': None}, u'uri': u'https://github.com/TomasTomecek/docker-hello-world.git', u'dockerfile_path': None}, u'target_registries': None, u'parent_registry': None}
2015-07-01 18:17:15,127 - atomic_reactor.source - DEBUG - workdir is '/tmp/tmpTF3oGV'
2015-07-01 18:17:15,127 - atomic_reactor.source - DEBUG - source path is '/tmp/tmpTF3oGV/source'
2015-07-01 18:17:15,135 - atomic_reactor.util - INFO - cloning git repo 'https://github.com/TomasTomecek/docker-hello-world.git'
2015-07-01 18:17:15,136 - atomic_reactor.util - DEBUG - url = 'https://github.com/TomasTomecek/docker-hello-world.git', dir = '/tmp/tmpTF3oGV/source', commit = 'master'
2015-07-01 18:17:15,136 - atomic_reactor.util - DEBUG - cloning single branch '[u'git', u'clone', u'-b', u'master', u'--single-branch', u'https://github.com/TomasTomecek/docker-hello-world.git', '/tmp/tmpTF3oGV/source']'
Cloning into '/tmp/tmpTF3oGV/source'...
2015-07-01 18:17:15,670 - atomic_reactor.util - DEBUG - getting SHA-1 of provided ref '[u'git', u'rev-parse', u'HEAD']'
2015-07-01 18:17:15,678 - atomic_reactor.util - INFO - commit ID = 1a595e502afba8942fe30caa09df7d442089db11
2015-07-01 18:17:15,679 - atomic_reactor.util - INFO - searching for dockerfile in '/tmp/tmpTF3oGV/source' (local path None)
2015-07-01 18:17:15,679 - atomic_reactor.util - DEBUG - abs path = '/tmp/tmpTF3oGV/source', local path = 'None'
2015-07-01 18:17:15,679 - atomic_reactor.util - DEBUG - Dockerfile found: '/tmp/tmpTF3oGV/source/Dockerfile'
2015-07-01 18:17:15,681 - atomic_reactor.build - DEBUG - base image specified in dockerfile = 'fedora:latest'
2015-07-01 18:17:15,681 - atomic_reactor.build - INFO - pulling base image 'fedora:latest' from registry 'None'
2015-07-01 18:17:15,682 - atomic_reactor.core - INFO - pulling image 'fedora:latest' from registry
2015-07-01 18:17:15,682 - atomic_reactor.core - DEBUG - image = 'fedora:latest', insecure = 'False'
Traceback (most recent call last):
  File "/usr/bin/atomic-reactor", line 9, in <module>
    load_entry_point('atomic-reactor==1.3.7', 'console_scripts', 'atomic-reactor')()
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/cli/main.py", line 284, in run
    cli.run()
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/cli/main.py", line 267, in run
    args.func(args)
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/cli/main.py", line 89, in cli_inside_build
    build_inside(input=args.input, input_args=args.input_arg, substitutions=args.substitute)
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/inner.py", line 336, in build_inside
    build_result = dbw.build_docker_image()
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/inner.py", line 241, in build_docker_image
    insecure=self.parent_registry_insecure)
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/build.py", line 126, in pull_base_image
    base_image = self.tasker.pull_image(base_image_with_registry, insecure=insecure)
  File "/usr/lib/python2.7/site-packages/atomic_reactor-1.3.7-py2.7.egg/atomic_reactor/core.py", line 369, in pull_image
    logs_gen = self.d.pull(image.to_str(tag=False), tag=image.tag, insecure_registry=insecure, stream=True)
  File "/usr/lib/python2.7/site-packages/docker/client.py", line 772, in pull
    headers=headers, stream=stream, timeout=None)
  File "/usr/lib/python2.7/site-packages/docker/client.py", line 100, in _post
    return self.post(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 508, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 465, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', error(13, 'Permission denied'))
```

Which then leads to the AVC denials in `/var/log/audit/audit.log`:
```
type=USER_AVC msg=audit(1435775358.150:464): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='avc:  received setenforce notice (enforcing=1)  exe="/usr/lib/systemd/systemd" sauid=0 hostname=? addr=? terminal=?'
type=ANOM_PROMISCUOUS msg=audit(1435775358.181:465): dev=veth8a3cae9 prom=256 old_prom=0 auid=4294967295 uid=0 gid=0 ses=4294967295
type=AVC msg=audit(1435775358.211:466): avc:  denied  { relabelfrom } for  pid=24493 comm="docker" name="/" dev="mqueue" ino=49852 scontext=system_u:system_r:docker_t:s0 tcontext=system_u:object_r:docker_tmpfs_t:s0 tclass=dir permissive=1
type=AVC msg=audit(1435775358.213:467): avc:  denied  { relabelfrom } for  pid=24493 comm="docker" name="tmpjakJfY" dev="vda1" ino=128854 scontext=system_u:system_r:docker_t:s0 tcontext=unconfined_u:object_r:user_tmp_t:s0 tclass=dir permissive=1
type=AVC msg=audit(1435775358.213:468): avc:  denied  { relabelfrom } for  pid=24493 comm="docker" name="build.json" dev="vda1" ino=128855 scontext=system_u:system_r:docker_t:s0 tcontext=unconfined_u:object_r:user_tmp_t:s0 tclass=file permissive=1
type=AVC msg=audit(1435775360.975:469): avc:  denied  { connectto } for  pid=24493 comm="atomic-reactor" path="/run/docker.sock" scontext=system_u:system_r:svirt_lxc_net_t:s0:c376,c586 tcontext=system_u:system_r:docker_t:s0 tclass=unix_stream_socket permissive=0
type=ANOM_PROMISCUOUS msg=audit(1435775361.095:470): dev=veth8a3cae9 prom=0 old_prom=256 auid=4294967295 uid=0 gid=0 ses=4294967295
type=SERVICE_STOP msg=audit(1435775391.361:471): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=systemd-machined comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
```

**Comment by TomasTomecek** [comment]
SELinux support is on our roadmap, unfortunately it has low prio. This may be actually dupe of https://github.com/DBuildService/atomic-reactor/issues/132

**Comment by maxamillion** [comment]
Just to note, the inability to run atomic-reactor (and ultimately OSBS) will block the [Fedora Docker Layered Image Build Service](https://fedoraproject.org/wiki/Changes/Layered_Docker_Image_Build_Service) change. The Fedora Infrastructure Team does not run production services that are not selinux enforcing.

**Comment by jpopelka** [comment]
We can ship a selinux module with atomic-reactor, the question is whether it's a good idea from security standpoint.

@rhatdan we need some help here. Do you think it'd be a big problem if we packaged a selinux module with atomic-reactor as described in #132 ?

> type=AVC msg=audit(1435775360.975:469): avc: denied { connectto } for pid=24493 comm="atomic-reactor" path="/run/docker.sock" scontext=system_u:system_r:svirt_lxc_net_t:s0:c376,c586 tcontext=system_u:system_r:docker_t:s0 tclass=unix_stream_socket permissive=0

This is the same as in #132 and I think it's caused by atomic-reactor running in container trying to connect to /run/docker.sock on host.

> type=AVC msg=audit(1435775358.213:468): avc: denied { relabelfrom } for pid=24493 comm="docker" name="build.json" dev="vda1" ino=128855 scontext=system_u:system_r:docker_t:s0 tcontext=unconfined_u:object_r:user_tmp_t:s0 tclass=file permissive=1

Not sure what can trigger this, any idea @TomasTomecek ?

**Comment by rhatdan** [comment]
You need an updated version of docker for the second AVC.  docker-1.6.2 and docker-1.7 should have docker-selinux installed which makes docker_t an unconfined process.

The first one is about a confined non privileged container trying to contact the docker daemon.  This is not something we want to allow. you could run the client without SELinux support

docker run --security-opt=label:disabled ...

Or run the docker daemon without SELinux support (Remove --selinux-enabled) from the /etc/sysconfig/docker.  


If the only containers you are going to run need to talk to docker, they basically have root access on the host system.  No security is going to block this.

**Comment by jpopelka** [comment]
Thanks.

If there's no chance that the policy can get into upstream selinux-policy, then shipping a selinux module with atomic-reactor looks like less evil than running docker daemon / container without SELinux at all. Or am I missing anything ?

Anyway I don't see a way how to influence how the docker daemon is run on build machine, neither do I see anything like --security-opt in [docker-py:create_container()](https://docker-py.readthedocs.org/en/latest/api/#create_container)

**Comment by rhatdan** [comment]
@jpopelka What policy do you want?  That A container process can communicate with the docker daemon?  We are not going to do this.  Perhaps we could do this is docker had any kind of RBAC at all.
Currently if I can talk to the docker daemon I can execute the following command

docker run -ti -v /:/host --privileged fedora chroot /host

And have full root running as an unconfined process on your system. 

Bottom line allowing the processes inside of the container access to /var/run/docker.sock, is giving them full root on the host.

**Comment by cgwalters** [comment]
Isn't this just that all use cases for containers that talk to Docker have to run with `--privileged` already?

**Comment by cgwalters** [comment]
For example, there's active work going on to containerize Kubernetes and OpenShift v3, which definitely requires `--privileged`, but the containers it spawns in response to API requests aren't themselves privileged and don't (by default) have access to the Docker socket.

**Comment by rhatdan** [comment]
Correct.

**Comment by cgwalters** [comment]
So then, whatever is running the container for atomic-reactor should just use `--privileged`.  I think this might be in the Ansible code?

**Comment by cgwalters** [comment]
(And it would be important to verify that the *build* containers don't have access to the socket)

**Comment by jpopelka** [comment]
@rhatdan  I'm also seeing this AVC (docker-1.7.1-5.gitc8962ac.fc22) when the atomic-reactor running in build container tries to open build.json, where /run/share/ is mounted inside the container and used as a way to transport data between host and the container. Any ideas what can we do with this one ?

type=AVC msg=audit(1438017356.268:1926): avc:  denied  { open } for  pid=1623 comm="atomic-reactor" path="/run/share/build.json" dev="tmpfs" ino=721673611 scontext=system_u:system_r:svirt_lxc_net_t:s0:c851,c884 tcontext=unconfined_u:object_r:user_tmp_t:s0 tclass=file permissive=0

**Comment by rhatdan** [comment]
You need to mount the /run/share with either a :Z or a :z.

```
-v /run/share:/run/share:Z
```

**Comment by rhatdan** [comment]
This will cause docker to relabel the content either private to the container or shared amongst containers.

**Comment by TomasTomecek** [comment]
looks like that the fix for this should be composed of two things:

1. run the build container as `privileged`
2. bindmount build json inside build container with `Z`

Once docker comes with some access control policy, we can get rid of 1).

**Comment by jpopelka** [comment]
>   1. run the build container as privileged
>   2. bindmount build json inside build container with Z

(2) should be fixed with 7080809 (needs docker-py >= 1.3.0)

> Once docker comes with some access control policy, we can get rid of 1).

And do we really need some some access control policy ? I don't think that's feasible in Fedora-23 time-frame, is it @rhatdan ?

At the moment we have 2 (if I don't count 'here' method) build methods: hostdocker & privileged.
@TomasTomecek  Would that be a problem if the hostdocker method created a privileged container too (even without the above mentioned policy) ?

**Comment by jpopelka** [comment]
> So then, whatever is running the container for atomic-reactor should just use --privileged. I think this might be in the Ansible code?

atomic-reactor on the host runs the build containers and according to the build method (`hostdocker` or `privileged`) decides whether to use `--privileged`

> (And it would be important to verify that the build containers don't have access to the socket)

I don't think I understand. The `hostdocker` method mounts the socket inside the build container to use the host's Docker, so they need to have access to it. And the SELinux AVC we're trying to work around here is AFAIK about SELinux on host denying it (the access).

**Comment by TomasTomecek** [comment]
> > Once docker comes with some access control policy, we can get rid of 1).

> And do we really need some some access control policy ? I don't think that's feasible in Fedora-23 time-frame, is it @rhatdan ?

It will take months (years?) to get there and we can't do pretty much anything.


I would run all containers as `privileged`. I'm not sure which caps we need actually to get rid of the AVCs. Also, since we need to access docker, we are basically opening doors to whole system, so why bother with security. Once docker starts taking security seriously, we can take care of this more securely. In the meantime, there's pretty much nothing we can do about it.

OpenShift is doing the same thing: it launches all custom builds as `privileged`.
